Moving build, testing, and cleanup into a try/finally block.

not-josh opened the pull request:
Trying to kill pytest with SIGINT causes docker, or it's network bridges, to stay alive. I discovered this when one of the surviving network bridges killed my internet connection entirely. At other times, my computer's performance will drop from leftover docker containers running. Life is hard for the impatient!

While this will not keep someone mashing ctrl+c safe, I've found this small change reduces the possibility of leftover waste without a noticeable performance cost.

AndreLouisCaron commented:
Hi  @not-josh !

Thanks for contributing :-)

I'm a bit surprised by this change.  According to the Pytest fixtures documentation, it doesn't seem like exceptions are propagated like this.  See [Fixture finalization / executing teardown code](https://docs.pytest.org/en/latest/fixture.html#fixture-finalization-executing-teardown-code).

not-josh commented:
Pytest won't run a fixture's teardown if the initial setup fails for _any_ reason. A try/finally block will still pass the exception upward while ensuring it's teardown code is still ran. 

Alternatively (and likely cleaner), this can also be done utilizing the **[addfinalizer](https://docs.pytest.org/en/latest/fixture.html#fixture-finalization-executing-teardown-code)** function of the pytest fixture subrequest. In the example below, I've added the request as a parameter, moved the teardown code into it's own function, and added that function the request using addfinalizer. What do you think?

```
@pytest.fixture(scope='session')
def docker_services(
    request, docker_compose_file, docker_allow_fallback, docker_compose_project_name
):
    """Ensure all Docker-based services are up and running."""

    def _cleanup():
        # Clean up.
        docker_compose.execute('down -v')

    docker_compose = DockerComposeExecutor(
        docker_compose_file, docker_compose_project_name
    )

    # If we allowed to run without Docker, check it's presence
    if docker_allow_fallback is True:
        try:
            execute('docker ps')
        except Exception:
            # Run against localhost
            yield Services(docker_compose, docker_allow_fallback=True)
            return

    # Spawn containers.
    docker_compose.execute('up --build -d')
    
    # If failure happens after this point, run cleanup
    request.addfinalizer(_cleanup)

    # Let test(s) run.
    yield Services(docker_compose)
```

not-josh commented:
Hello @butla! 

Raising an exception during or after `docker_compose.execute('up --build -d')` and before our `yield` will result in your docker images staying up and running **without** the use of a finalizer. The docker images _will_ clean themselves up with a defined finalizer telling them to. 

Part of this is because `docker-compose` is being ran with the `-d` (detached) flag. Normally tossing `up` a SIGINT will result in a cleanup, however it is expected for you to manually clean things up using `down` while running it detached. 

As for `up` running alongside `down`... The way `subprocess.check_output` behaves will ensure that `up` will not be ran in parallel to `down`. Once our **Python process** runs into a KeyboardInterrupt (or any other error) check_output will [kill the background process it's waiting on](https://github.com/python/cpython/blob/master/Lib/subprocess.py#L481) using [SIGKILL](https://github.com/python/cpython/blob/master/Lib/subprocess.py#L1761), ensuring the process dies immediately _before_ raising whatever exception occurred. 

Running `docker-compose down` is also pretty safe -- even if our `up` command failed halfway through! `down` will only try to [remove things created by the Compose file](https://docs.docker.com/compose/reference/down/) and does not raise any errors if some (or all) images aren't running. It'll complain, but still no errors!

not-josh commented on src/pytest_docker/__init__.py:
```diff
@@ -140,22 +140,26 @@ def docker_compose_project_name():
     docker_compose.execute('up --build -d')
 
     # Let test(s) run.
     yield Services(docker_compose)
```
Agreed, will change.

not-josh commented on src/pytest_docker/__init__.py:
```diff
@@ -140,22 +140,26 @@ def docker_compose_project_name():
         docker_compose_file, docker_compose_project_name
     )
 
+    # If failure happens beyond this point, run cleanup
```
Code does look a little spread out. I've removed two lines of white space, but kept the newlines between comments.

not-josh commented:
Also, I've fixed the pep8 errors with the tests, however they're still broken due to unrelated changes. :(

butla commented:
Issue, leading to the code fix and some thoughts - #35

augi commented:
Is this still valid, please?
